go/parquet/internal/encoding/byte_stream_split.go

```diff
@@ -0,0 +1,367 @@
+}
+
+func (enc *ByteStreamSplitFloat64Encoder) FlushValues() (Buffer, error) {
+	in, err := enc.PlainFloat64Encoder.FlushValues()
+	if err != nil {
+		return nil, err
+	}
+
+	if enc.flushBuffer == nil {
+		enc.flushBuffer = NewPooledBufferWriter(in.Len())
+	}
+
+	enc.flushBuffer.buf.Resize(in.Len())
+	encodeByteStreamSplitWidth8(enc.flushBuffer.Bytes(), in.Bytes())
+	return enc.flushBuffer.Finish(), nil
+}
+
+func (enc *ByteStreamSplitFloat64Encoder) Release() {
+	enc.PlainFloat64Encoder.Release()
+	releaseBufferToPool(enc.flushBuffer)
+	enc.flushBuffer = nil
+}
+
+// ByteStreamSplitInt32Encoder writes the underlying bytes of the Int32
+// into interlaced streams as defined by the BYTE_STREAM_SPLIT encoding
+type ByteStreamSplitInt32Encoder struct {
+	PlainInt32Encoder
+	flushBuffer *PooledBufferWriter
+}
+
+func (enc *ByteStreamSplitInt32Encoder) FlushValues() (Buffer, error) {
+	in, err := enc.PlainInt32Encoder.FlushValues()
+	if err != nil {
+		return nil, err
+	}
+
+	if enc.flushBuffer == nil {
+		enc.flushBuffer = NewPooledBufferWriter(in.Len())
+	}
+
+	enc.flushBuffer.buf.Resize(in.Len())
+	encodeByteStreamSplitWidth4(enc.flushBuffer.Bytes(), in.Bytes())
+	return enc.flushBuffer.Finish(), nil
+}
+
+func (enc *ByteStreamSplitInt32Encoder) Release() {
+	enc.PlainInt32Encoder.Release()
+	releaseBufferToPool(enc.flushBuffer)
+	enc.flushBuffer = nil
+}
+
+// ByteStreamSplitInt64Encoder writes the underlying bytes of the Int64
+// into interlaced streams as defined by the BYTE_STREAM_SPLIT encoding
+type ByteStreamSplitInt64Encoder struct {
+	PlainInt64Encoder
+	flushBuffer *PooledBufferWriter
+}
+
+func (enc *ByteStreamSplitInt64Encoder) FlushValues() (Buffer, error) {
+	in, err := enc.PlainInt64Encoder.FlushValues()
+	if err != nil {
+		return nil, err
+	}
+
+	if enc.flushBuffer == nil {
+		enc.flushBuffer = NewPooledBufferWriter(in.Len())
+	}
+
+	enc.flushBuffer.buf.Resize(in.Len())
+	encodeByteStreamSplitWidth8(enc.flushBuffer.Bytes(), in.Bytes())
+	return enc.flushBuffer.Finish(), nil
+}
+
+func (enc *ByteStreamSplitInt64Encoder) Release() {
+	enc.PlainInt64Encoder.Release()
+	releaseBufferToPool(enc.flushBuffer)
+	enc.flushBuffer = nil
+}
+
+type ByteStreamSplitFloat32Decoder = ByteStreamSplitDecoder[float32]
+type ByteStreamSplitFloat64Decoder = ByteStreamSplitDecoder[float64]
+type ByteStreamSplitInt32Decoder = ByteStreamSplitDecoder[int32]
+type ByteStreamSplitInt64Decoder = ByteStreamSplitDecoder[int64]
+
+type ByteStreamSplitDecoder[T float32 | float64 | int32 | int64] struct {
+	decoder
+	stride int
+}
+
+func (dec *ByteStreamSplitDecoder[T]) Type() parquet.Type {
+	switch any(dec).(type) {
+	case *ByteStreamSplitDecoder[float32]:
+		return parquet.Types.Float
+	case *ByteStreamSplitDecoder[float64]:
+		return parquet.Types.Double
+	case *ByteStreamSplitDecoder[int32]:
+		return parquet.Types.Int32
+	case *ByteStreamSplitDecoder[int64]:
+		return parquet.Types.Int64
+	default:
+		return parquet.Types.Undefined
+	}
```

Comment:
since we should never hit this (because of the constraint on the type) let's make this a panic instead